## Extra: Tripulación del Titanic

Si llegaste hasta acá y todavía tenés tiempo y ganas, te proponemos que explores el dataset de Supervivientes del Titanic. Queremos saber si los sobrevivientes del RMS Titanic murieron *al azar* o si los sobrevivientes tenían alguna característica distinta de las víctimas de la tragedia. Tenemos disponibles algunos datos sobre pasajeros/as del Titanic, que pueden descargar [acá](https://www.kaggle.com/c/titanic/) (solo el archivo `train.csv`). La propuesta es analizarlos para intentar encontrar algunas características que separen a los pasajeros que sobrevivieron de los que no.

### Análisis Exploratorio de Datos

1. Cargar los datos desde el archivo csv.
1. Investigar qué forma tienen los datos utilizando las funciones exploratorias que aprendimos. ¿Qué información tenemos en los datos?¿Qué representa cada columna?¿Cuántos pasajeros están incluidos en este Dataset?
1. ¿Faltan datos? ¿Se te ocurre por qué? ¿Qué harías con ellos?
1. ¿Te parece que todas las columnas son informativas o borrarías alguna?

**Para pensar**: ¿te parece que la supervivencia (o no) fue un proceso completamente *al azar* o existe algún mecanismo generador de estos datos?¿Qué nos enseñó la famosa película de David Cameron?

https://medium.com/@diolmedo/participando-en-la-competencia-titanic-machine-learning-from-disaster-en-la-plataforma-kaggle-47f01c838af2

In [35]:
## Importamos Pandas as pd y, 
## Matplotlib.pyplot as plt:

import pandas as pd 
import matplotlib.pyplot as plt


In [36]:
## Leemos el archivo csv con la función pd read y le asignamos el nombre df (DataFrame) a la variable.

df_train = pd.read_csv('./Titanic_Dataset/train.csv')
df_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

2. ¿Cuántas columnas (features) tiene?¿Cuáles son sus nombres?¿Y cuántas filas (instancias)?.

In [15]:
df_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

Originalmente tiene 12 columnas y 891 filas.

3. ¿Cuántos valores faltantes hay en cada columna?

In [31]:
## Itero sobre las columnas y printeo sobre 

for column in df_train:
    print(f'En la columna \"{column}\" hay', df_train[f'{column}'].isna().sum(), 'datos faltantes.')

# df_train.isna().sum()             METODO DE SAMEK QUE TAMBIÉN FUNCIONA. NASHE


En la columna "PassengerId" hay 0 datos faltantes.
En la columna "Survived" hay 0 datos faltantes.
En la columna "Pclass" hay 0 datos faltantes.
En la columna "Name" hay 0 datos faltantes.
En la columna "Sex" hay 0 datos faltantes.
En la columna "Age" hay 177 datos faltantes.
En la columna "SibSp" hay 0 datos faltantes.
En la columna "Parch" hay 0 datos faltantes.
En la columna "Ticket" hay 0 datos faltantes.
En la columna "Fare" hay 0 datos faltantes.
En la columna "Cabin" hay 687 datos faltantes.
En la columna "Embarked" hay 2 datos faltantes.


In [ ]:
df_train.isna(axis = 1).sum() 

4. ¿Cuál o cuáles son las columnas con más valores faltantes?

Las columnas con mas valores faltantes son:
- 1. Cabin (687).
- 2. Age (177).
- 3. Embarked (2).

5. Hacerse alguna pregunta acerca del dataset e intentar responderla. Por ejemplo, ¿cuál es la persona de más edad?

- Reemplazar los valores numericos y alfa-numericos a palabras completas. 
<br>
<br>
    -   Por ejemplo: 0 = False, 1 = True, etc.


In [39]:
## Podemos ver que en la columna Survived el numero 0 representa false y el numero 1 representa true, entonces procedemos a reemplazarlos en todas las filas de la misma.

df_train["Survived"] = df_train["Survived"].astype(bool)
df_train


# Southampton
# Cherbourg
# QueensTown



PassengerId  Survived  Pclass  \
0              1     False       3   
1              2      True       1   
2              3      True       3   
3              4      True       1   
4              5     False       3   
..           ...       ...     ...   
886          887     False       2   
887          888      True       1   
888          889     False       3   
889          890      True       1   
890          891     False       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

## Trabajando con datos faltantes

Antes de realizar cada una de las siguientes consignas, recordá copiar el dataset en una nueva variable, así mantemos el dataset original sin modificar.

1. Descartar aquellas filas que tengan algún valor faltante. ¿Cuántas instancias quedan en el dataset?

In [66]:
df_train2 = df_train.dropna().reset_index()
df_train2

index  PassengerId  Survived  Pclass  \
0        1            2      True       1   
1        3            4      True       1   
2        6            7     False       1   
3       10           11      True       3   
4       11           12      True       1   
..     ...          ...       ...     ...   
178    871          872      True       1   
179    872          873     False       1   
180    879          880      True       1   
181    887          888      True       1   
182    889          890      True       1   

                                                  Name     Sex   Age  SibSp  \
0    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
1         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
2                              McCarthy, Mr. Timothy J    male  54.0      0   
3                      Sandstrom, Miss. Marguerite Rut  female   4.0      1   
4                             Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
178   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
179                           Carlsson, Mr. Frans Olof    male  33.0      0   
180      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
181                       Graham, Miss. Margaret Edith  female  19.0      0   
182                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
0        0  PC 17599  71.2833          C85        C  
1        0    113803  53.1000         C123        S  
2        0     17463  51.8625          E46        S  
3        1   PP 9549  16.7000           G6        S  
4        0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
178      1     11751  52.5542          D35        S  
179      0       695   5.0000  B51 B53 B55        S  
180      1     11767  83.1583          C50        C  
181      0    112053  30.0000          B42        S  
182      0    111369  30.0000         C148        C  

[183 rows x 13 columns]

In [71]:
df_train2 = df_train2.drop("index", axis=1)

In [72]:
df_train2

PassengerId  Survived  Pclass  \
0              2      True       1   
1              4      True       1   
2              7     False       1   
3             11      True       3   
4             12      True       1   
..           ...       ...     ...   
178          872      True       1   
179          873     False       1   
180          880      True       1   
181          888      True       1   
182          890      True       1   

                                                  Name     Sex   Age  SibSp  \
0    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
1         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
2                              McCarthy, Mr. Timothy J    male  54.0      0   
3                      Sandstrom, Miss. Marguerite Rut  female   4.0      1   
4                             Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
178   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
179                           Carlsson, Mr. Frans Olof    male  33.0      0   
180      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
181                       Graham, Miss. Margaret Edith  female  19.0      0   
182                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
0        0  PC 17599  71.2833          C85        C  
1        0    113803  53.1000         C123        S  
2        0     17463  51.8625          E46        S  
3        1   PP 9549  16.7000           G6        S  
4        0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
178      1     11751  52.5542          D35        S  
179      0       695   5.0000  B51 B53 B55        S  
180      1     11767  83.1583          C50        C  
181      0    112053  30.0000          B42        S  
182      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

Si eliminamos los datos faltantes nos quedarían 12 columnas y 183 filas.

2. Descartar aquellas columnas que tengan algún valor faltante. ¿Cuántos y cuáles atributos quedaron?

In [77]:
## Copiamos el df original con todos sus datos faltantes

df3 = df_train.copy()

Pueden observar que descartar sin ningún criterio puede hacer que perdamos muchos datos.

Tratemos de mejorar un poco los criterios con los que descartamos.

3. Descartar aquellas filas que tengan más de tres valores faltantes. ¿Cuántas instancias quedaron?

In [108]:
## Copiamos el df original con todos sus datos faltantes

df3 = df_train.copy()
df3 = df3.dropna(axis = 0, inplace=True, thresh = 3)


In [109]:
df3

4. Descartar aquellas columnas que tengan más del 50% de valores faltantes. ¿Cuántos y cuáles atributos quedaron? Luego, descartar las instancias que tengan algún valor faltante.

## Imputación de Valores Faltantes

1. Calcular el valor medio, moda, mediana y la desviación estándar de la columna `price`.

2. Imputar los valores faltantes en la columna `Age` usando su valor medio.

3. ¿Cambió el valor medio de la columna?¿Y su desviación estándar?¿Y los otros estadísticos que calculamos?¿Por qué?

4. Volver a correr las celdas anteriores, pero completando los valores faltantes con la mediana y luego la moda.

5. Trabajar con un dataset sin visualizar es muy aburrido y, diríamos, casi frustrante. La próxima clase ya nos meteremos de lleno en visualización. Mientras tanto, intenten hacer un histograma de los puntajes asignados y, luego, de los precios.